File: windows-build-20260529-stack-box-direct-real-bepu/cooked/code/gameplay/MenuDefinitionProviderResolver.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "MenuDefinitionProviderResolver.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_type.hpp"
#include "IMenuDefinitionProvider.hpp"
#include "MenuDefinition.hpp"
#include "MenuDefinitionProviderResolver.hpp"
#include "runtime/array.hpp"
#include "IScriptTypeResolver.hpp"
#include "runtime/array.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_type.hpp"

MenuDefinitionProviderResolver::MenuDefinitionProviderResolver(IScriptTypeResolver* scriptTypeResolver) : ScriptTypeResolver()
{
this->ScriptTypeResolver = scriptTypeResolver;
}

::IMenuDefinitionProvider* MenuDefinitionProviderResolver::Resolve(std::string providerTypeName)
{
    if (String::IsNullOrWhiteSpace(providerTypeName))
    {
throw ([&]() {
auto __ctor_arg_00000036 = "Provider type name must be provided.";
auto __ctor_arg_00000037 = "providerTypeName";
return new ArgumentException(__ctor_arg_00000036, __ctor_arg_00000037);
})();
    }
Type *providerType = Type::GetType(providerTypeName, false);
    if (providerType == nullptr && this->ScriptTypeResolver != nullptr)
    {
providerType = this->ScriptTypeResolver->Resolve(providerTypeName);
    }
    if (providerType == nullptr)
    {
throw new InvalidOperationException(std::string("Menu provider type '") + providerTypeName + std::string("' could not be resolved."));
    }
    if (!he_cpp_type_of<IMenuDefinitionProvider>("IMenuDefinitionProvider")->IsAssignableFrom(providerType))
    {
throw new InvalidOperationException(std::string("Menu provider type '") + providerTypeName + std::string("' must implement ") + "IMenuDefinitionProvider" + std::string("."));
    }
ConstructorInfo *constructor = providerType->GetConstructor(Type::EmptyTypes);
    if (constructor == nullptr || !constructor->get_IsPublic())
    {
throw new InvalidOperationException(std::string("Menu provider type '") + providerTypeName + std::string("' must expose a public parameterless constructor."));
    }
const void *instance = Activator::CreateInstance(providerType);
::IMenuDefinitionProvider *provider = he_cpp_try_cast<IMenuDefinitionProvider>(instance);
    if (provider == nullptr)
    {
throw new InvalidOperationException(std::string("Menu provider type '") + providerTypeName + std::string("' could not be instantiated."));
    }
return provider;}

File: assets/codebase/menu.authoring/MenuDefinitionProviderResolver.cs
namespace city.menu {
    /// <summary>
    /// Resolves city menu-definition providers from assembly-qualified type names persisted in scenes.
    /// </summary>
    public sealed class MenuDefinitionProviderResolver {
        /// <summary>
        /// Optional script type resolver used for module-qualified provider type names.
        /// </summary>
        readonly IScriptTypeResolver ScriptTypeResolver;

        /// <summary>
        /// Initializes a new city menu-definition provider resolver.
        /// </summary>
        /// <param name="scriptTypeResolver">Optional shared script type resolver used for loaded gameplay modules.</param>
        public MenuDefinitionProviderResolver(IScriptTypeResolver scriptTypeResolver = null) {
            ScriptTypeResolver = scriptTypeResolver;
        }

        /// <summary>
        /// Instantiates one city menu-definition provider from an assembly-qualified type name.
        /// </summary>
        /// <param name="providerTypeName">Assembly-qualified type name of the provider.</param>
        /// <returns>Instantiated provider implementation.</returns>
        [NativeOwnedReturn]
        public IMenuDefinitionProvider Resolve(string providerTypeName) {
            if (string.IsNullOrWhiteSpace(providerTypeName)) {
                throw new ArgumentException("Provider type name must be provided.", nameof(providerTypeName));
            }

#if HELENGINE_CODEGEN_DISABLE_MENU_REFLECTION
            throw new InvalidOperationException("Menu definition provider reflection is not available in generated native builds.");
#else
            Type providerType = Type.GetType(providerTypeName, false);
            if (providerType == null && ScriptTypeResolver != null) {
                providerType = ScriptTypeResolver.Resolve(providerTypeName);
            }
            if (providerType == null) {
                throw new InvalidOperationException($"Menu provider type '{providerTypeName}' could not be resolved.");
            }
            if (!typeof(IMenuDefinitionProvider).IsAssignableFrom(providerType)) {
                throw new InvalidOperationException($"Menu provider type '{providerTypeName}' must implement {nameof(IMenuDefinitionProvider)}.");
            }

            var constructor = providerType.GetConstructor(Type.EmptyTypes);
            if (constructor == null || !constructor.IsPublic) {
                throw new InvalidOperationException($"Menu provider type '{providerTypeName}' must expose a public parameterless constructor.");
            }

            object instance = Activator.CreateInstance(providerType);
            IMenuDefinitionProvider provider = instance as IMenuDefinitionProvider;
            if (provider == null) {
                throw new InvalidOperationException($"Menu provider type '{providerTypeName}' could not be instantiated.");
            }

            return provider;
#endif
        }
    }
}
